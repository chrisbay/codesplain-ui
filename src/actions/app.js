import axios from 'axios';
import cookie from 'react-cookie';

export const CLEAR_UNSAVED_CHANGES = 'CLEAR_UNSAVED_CHANGES';
export const EDIT_ANNOTATION = 'EDIT_ANNOTATION';
export const PARSE_SNIPPET = 'PARSE_SNIPPET';
export const RESTORE_STATE = 'RESTORE_STATE';
export const RESET_RULE_FILTERS = 'RESET_RULE_FILTERS';
export const SAVE_ANNOTATION = 'SAVE_ANNOTATION';
export const SET_AST = 'SET_AST';
export const SET_RULE_FILTERS = 'SET_RULE_FILTERS';
export const SET_SNIPPET_CONTENTS = 'SET_SNIPPET_CONTENTS';
export const SET_SNIPPET_LANGUAGE = 'SET_SNIPPET_LANGUAGE';
export const SET_SNIPPET_TITLE = 'SET_SNIPPET_TITLE';
export const TOGGLE_EDITING_STATE = 'TOGGLE_EDITING_STATE';
export const SAVE_STATE = 'SAVE_STATE';
export const SAVE_STATE_STARTED = 'SAVE_STATE_STARTED';
export const SAVE_STATE_SUCCEEDED = 'SAVE_STATE_SUCCEEDED';
export const SAVE_STATE_FAILED = 'SAVE_STATE_FAILED';
export const SELECT_ALL_FILTERS = 'SELECT_ALL_FILTERS';
export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const UPDATE_USER_SNIPPETS_STARTED = 'UPDATE_USER_SNIPPETS_STARTED';
export const UPDATE_USER_SNIPPETS_SUCCEEDED = 'UPDATE_USER_SNIPPETS_SUCCEEDED';
export const UPDATE_USER_SNIPPETS_FAILED = 'UPDATE_USER_SNIPPETS_FAILED';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';

const API_URL = process.env.REACT_APP_API_URL;

export const setSnippetContents = (snippet) => ({
  type: SET_SNIPPET_CONTENTS,
  payload: snippet,
});

export const setRuleFilters = (filters) => ({
  type: SET_RULE_FILTERS,
  payload: filters,
});

export const selectAllFilters = () => ({
  type: SELECT_ALL_FILTERS,
});

export const resetFilters = () => ({
  type: RESET_RULE_FILTERS,
});

export const setAST = (AST) => ({
  type: SET_AST,
  payload: AST,
});

export const toggleEditState = () => ({
  type: TOGGLE_EDITING_STATE,
});

export const saveAnnotation = (annotationData) => ({
  type: SAVE_ANNOTATION,
  payload: annotationData,
});

export const restoreState = (savedState) => ({
  type: RESTORE_STATE,
  payload: savedState,
});

export const setSnippetTitle = (snippetTitle) => ({
  type: SET_SNIPPET_TITLE,
  payload: snippetTitle
})

export const setSnippetLanguage = (snippetLanguage) => ({
  type: SET_SNIPPET_LANGUAGE,
  payload: snippetLanguage
});

export const parseSnippet = (snippet) => ({
  type: PARSE_SNIPPET,
  meta: {
    WebWorker: true,
  },
  payload: {
    snippet,
  },
});

export const clearUnsavedChanges = () => ({
  type: CLEAR_UNSAVED_CHANGES,
})

export const saveStateStarted = () => ({
  type: SAVE_STATE_STARTED,
});

export const saveStateSucceeded = () => ({
  type: SAVE_STATE_SUCCEEDED,
});

export const saveStateFailed = () => ({
  type: SAVE_STATE_FAILED,
});

export const setUserSnippets = (snippetMeta) => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const updateUserSnippetsStarted = () => ({
  type: UPDATE_USER_SNIPPETS_STARTED,
});

export const updateUserSnippetsSucceeded = () => ({
  type: UPDATE_USER_SNIPPETS_SUCCEEDED,
});

export const updateUserSnippetsFailed = () => ({
  type: UPDATE_USER_SNIPPETS_FAILED,
});

export const updateUserSnippets = () => {
  return (dispatch) => {
    // Load requisite cookies, return Promise if they aren't present
    const token    = cookie.load('token');
    const username = cookie.load('username');
    if (!token || !username) {
      return Promise.resolve();
    }

    // Fetch the user's snippet meta data and save it
    const headers = {
      Accept: 'application/json',
      Authorization: `token ${token}`,
    };
    dispatch(updateUserSnippetsStarted());
    return axios.get(`${API_URL}/users/${username}/snippets`, { headers })
      .then(
        res => {
          dispatch(setUserSnippets(res));
          dispath(updateUserSnippetsSucceeded());
        },
        err => dispatch(updateUserSnippetsFailed)
      );
  };
};

export const saveState = (id) => {
  return (dispatch, getState) => {
    // If the snippet has not been saved, then saving an annotation shouldn't
    // trigger a POST request to save the snippet (and its annotations)
    const username = cookie.load('username');
    if (id === undefined || !username) {
      // Return a Promise object to make this action "thenable"
      return Promise.resolve();
    }
    dispatch(saveStateStarted());
    // Get the app state and stringify it
    const stateString = JSON.stringify(getState().app);
    return axios.post(`${API_URL}/users/${username}/snippets/${id}`, { json: stateString, })
      .then((res) => {
        dispatch(saveStateSucceeded());
      }, (err) => {
        dispatch(saveStateFailed());
      });
  };
};
