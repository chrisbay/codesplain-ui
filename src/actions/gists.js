import {
  parseSnippet,
  setSnippetContents,
} from './app';
import {
  fetchGist as makeFetchGistRequest,
  fetchGists as makeFetchGistsRequest,
} from '../util/requests';

export const SET_GISTS = 'SET_GISTS';

export const setGists = gists => ({
  type: SET_GISTS,
  payload: gists,
});

export const fetchGist = url => dispatch => (
  makeFetchGistRequest(url)
    .then((contents) => {
      dispatch(setSnippetContents(contents));
      dispatch(parseSnippet(contents));
    })
);

export const fetchGists = () => (dispatch, getState) => {
  const { token } = getState().user;
  return makeFetchGistsRequest(token)
    .then((gists) => { dispatch(setGists(gists)); });
};
