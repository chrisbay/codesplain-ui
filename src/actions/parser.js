export const SET_PARSER_STATE = 'SET_PARSER_STATE';
export const LOAD_PARSER = 'LOAD_PARSER'
export const LOADING = 'LOADING'
export const LOADED = 'LOADED'

export const setParserState = (state) => ({
  type: SET_PARSER_STATE,
  payload: state
})

export const loadParser = (parser) => ({
  type: LOAD_PARSER,
  payload: parser
})
