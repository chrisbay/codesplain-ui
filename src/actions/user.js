import axios from 'axios';
import _ from 'lodash';

import { makeSaveEndpointUrl, sanitizeSnippetList } from '../util/requests';

const API_URL = process.env.REACT_APP_API_URL;

// Util func to check for 'NoSuchKey' responses from S3
const noSuchKey = data => (
  typeof data === 'string' && data.includes('NoSuchKey')
);

export const ADD_ORG = 'ADD_ORG';
export const ADD_ORGANIZATIONS = 'ADD_ORGANIZATIONS';
export const CLEAR_USER_CREDENTIALS = 'CLEAR_USER_CREDENTIALS';
export const RESTORE_USER_CREDENTIALS = 'RESTORE_USER_CREDENTIALS';
export const SAVE_ACCESS_TOKEN = 'SAVE_ACCESS_TOKEN';
export const SAVE_USERNAME = 'SAVE_USERNAME';
export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const SET_SNIPPET_LISTS = 'SET_SNIPPET_LISTS';
export const SWITCH_ORG = 'SWITCH_ORG';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';
export const UPDATE_USER_SNIPPETS_FAILED = 'UPDATE_USER_SNIPPETS_FAILED';
export const UPDATE_USER_SNIPPETS_STARTED = 'UPDATE_USER_SNIPPETS_STARTED';
export const UPDATE_USER_SNIPPETS_SUCCEEDED = 'UPDATE_USER_SNIPPETS_SUCCEEDED';

export const addOrg = org => ({
  type: ADD_ORG,
  payload: org,
});

export const addOrganizations = organizations => ({
  type: ADD_ORGANIZATIONS,
  payload: organizations,
});

export const clearUserCredentials = () => ({
  type: CLEAR_USER_CREDENTIALS,
});

export const saveAccessToken = accessToken => ({
  type: SAVE_ACCESS_TOKEN,
  payload: accessToken,
});

export const saveUsername = username => ({
  type: SAVE_USERNAME,
  payload: username,
});

export const setSnippetLists = snippetLists => ({
  type: SET_SNIPPET_LISTS,
  payload: snippetLists,
});

export const setUserSnippets = snippetMeta => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const switchOrg = org => ({
  type: SWITCH_ORG,
  payload: org,
});

export const updateUserSnippetsStarted = () => ({
  type: UPDATE_USER_SNIPPETS_STARTED,
});

export const updateUserSnippetsSucceeded = () => ({
  type: UPDATE_USER_SNIPPETS_SUCCEEDED,
});

export const updateUserSnippetsFailed = () => ({
  type: UPDATE_USER_SNIPPETS_FAILED,
});

export const updateUserSnippets = () => (dispatch, getState) => {
  // Load requisite cookies, return Promise if they aren't present
  const {
    user: {
      token,
      username,
    },
  } = getState();
  // Fetch the user's snippet meta data and save it
  const headers = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  dispatch(updateUserSnippetsStarted());
  return axios.get(makeSaveEndpointUrl(username), { headers })
    .then((res) => {
      // Jump to catch block if the user has no index.json file:
      if (noSuchKey(res.data)) {
        throw new Error(`index.json does not exist for ${username}!`);
      }
      dispatch(setUserSnippets(res.data));
      dispatch(updateUserSnippetsSucceeded());
    })
    .catch(() => {
      dispatch(updateUserSnippetsFailed());
    });
};

export const fetchSnippetLists = () => (dispatch, getState) => {
  const {
    token,
    orgs,
  } = getState().user;
  const headers = {
    Authorization: `token ${token}`,
  };
  const url = `${API_URL}/users?users=${orgs.join(',')}`;
  // Remove the username from the org list
  return axios({
    method: 'GET',
    url,
    headers,
  })
    .then(({ data: snippetLists }) => {
      const sanitizedSnippetLists = _.mapValues(snippetLists, sanitizeSnippetList);
      dispatch(setSnippetLists(sanitizedSnippetLists));
    });
};

export const fetchAccessToken = authCode => () => {
  const reqUrl = `${API_URL}/auth`;
  const reqBody = { code: authCode };
  return axios({
    method: 'POST',
    url: reqUrl,
    data: reqBody,
  });
};

export const fetchUserInfo = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user',
    headers: reqHeaders,
  });
};

export const restoreUserCredentials = (token, username) => ({
  type: RESTORE_USER_CREDENTIALS,
  payload: {
    token,
    username,
  },
});
