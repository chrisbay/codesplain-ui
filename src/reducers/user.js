import * as actions from '../actions/user';

const defaultState = {
  orgs: [],
  selectedOrg: null
}

const user = (state=defaultState, action) => {
  switch (action.type) {
    case actions.ADD_ORG: {
      const org = action.payload;
      if(state.orgs.includes(org)) {
        return state;
      } else {
        return {
          ...state,
          orgs: state.orgs.slice().concat(org)
        };
      }
    }
    case actions.SWITCH_ORG: {
      const org = action.payload;
      if (state.orgs.includes(org)) {
        return {
          ...state,
          selectedOrg: org
        };
      } else {
        return state;
      }
    }
    case actions.SET_USER_SNIPPETS: {
      return {
        ...state,
        userSnippets: action.payload,
      }
    }
    default: {
      return state;
    }
  }
};

export default user;
