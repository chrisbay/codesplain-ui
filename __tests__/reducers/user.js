import * as actions from '../../src/actions/user';
import reducer from '../../src/reducers/user';

describe('Reducer: User', () => {
  it('shoud have initial state', () => {
    const initial = {
      orgs: [],
      selectedOrg: null
    };
    expect(reducer(undefined, {})).toEqual(initial);
  });

  it('should handle ADD_ORG', () => {
    const org = 'galactic-federation';
    const action = {
      type: actions.ADD_ORG,
      payload: org
    }
    expect(reducer(undefined, action).orgs).toEqual(expect.arrayContaining([org]))
  });

  describe('SWITCH_ORG', () => {
    const state = {};
    const validOrg = 'galactic-federation';
    const invalidOrg = 'international-counsel-of-ricks';
    beforeEach(() => {
      state.orgs = [validOrg];
      state.selectedOrg = null
    });
    it('should set the org if the org exists', () => {
      const action = {
        type: actions.SWITCH_ORG,
        payload: validOrg
      };
      expect(reducer(state, action).selectedOrg).toEqual(validOrg);
    });
    it('shouldn\'t set the org if the org doesn\'t exists', () => {
      const action = {
        type: actions.SWITCH_ORG,
        payload: invalidOrg
      };
      expect(reducer(state, action).selectedOrg).toBe(null);
    });
  });

  it('should handle SET_USER_SNIPPETS', () => {
    const userSnippets = {
      'test_snippet_1': {
        snippetName: 'Test Snippet 1',
        language:    'Python 3',
        lastEdited:  '2017-04-05T12:52:20.099Z',
      },
      'test_snippet_2': {
        snippetName: 'Test Snippet 2',
        language:    'Python 3',
        lastEdited:  '2017-04-05T12:52:20.099Z',
      }
    };
    const action = {
      type: actions.SET_USER_SNIPPETS,
      payload: userSnippets,
    };
    const expected = { userSnippets };
    expect(reducer(undefined, action)).toEqual(expect.objectContaining(expected));
  });
});
