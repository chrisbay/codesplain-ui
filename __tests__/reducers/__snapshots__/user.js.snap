exports[`Reducer: User initialState matches snapshot 1`] = `
Object {
  "orgSnippets": Object {},
  "orgs": Array [],
  "selectedOrg": null,
  "token": "",
  "userSnippets": Object {},
  "username": "",
}
`;
